src/main/java/nebula/plugin/metrics/dispatcher/AbstractMetricsDispatcher.java

```diff
@@ -0,0 +1,273 @@
+                } catch (JsonProcessingException e) {
+                    logger.error("Unable to write JSON string value: " + e.getMessage(), e);
+                }
+            }
+        };
+
+        if (executeSynchronously) {
+            executeSynchronously(runnable);
+        } else {
+            queue(runnable);
+        }
+    }
+
+    /*
+     * Override this method to transform the Build object into a different Build representation. For
+     * example, when the Build format needs to be flattened out.
+     */
+    protected Object transformBuild(Build build) {
+        checkNotNull(build);
+        return build;
+    }
+
+    private void sanitizeProperties(Build build) {
+        Info info = build.getInfo();
+        if (info != null) {
+            build.setInfo(Info.sanitize(info, extension.getSanitizedProperties()));
+        }
+    }
+
+    @Override
+    public final void started(Project project) {
+        build.setProject(project);
+        indexBuildModel(false);
+    }
+
+    @Override
+    public final void duration(long startTime, long elapsedTime) {
+        build.setStartTime(startTime);
+        build.setElapsedTime(elapsedTime);
+    }
+
+    @Override
+    public final void environment(Info info) {
+        build.setInfo(info);
+    }
+
+    @Override
+    public final void result(Result result) {
+        build.setResult(result);
+    }
+
+    @Override
+    public final void event(String description, String type, long elapsedTime) {
+        build.addEvent(new Event(description, type, elapsedTime));
+    }
+
+    @Override
+    public final void task(Task task) {
+        build.addTask(task);
+    }
+
+    @Override
+    public final void test(Test test) {
+        build.addTest(test);
+    }
+
+    @Override
+    public final void logEvent(final LogEvent event) {
+        checkNotNull(event);
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                String json = renderEvent(event);
+                index(getLogCollectionName(), LOG_TYPE, json, Optional.<String>absent());
+            }
+        };
+        queue(runnable);
+    }
+
+    @Override
+    public final void logEvents(final Collection<LogEvent> events) {
+        checkNotNull(events);
+        checkArgument(!events.isEmpty(), "Empty events list found: %s", events);
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                List<String> jsons = Lists.newArrayListWithCapacity(events.size());
+                for (LogEvent event : events) {
+                    jsons.add(renderEvent(event));
+                }
+                bulkIndex(getLogCollectionName(), LOG_TYPE, jsons);
+            }
+        };
+        queue(runnable);
+    }
+
+    protected String renderEvent(LogEvent event) {
+        return event.toString();
+    }
+
+
+    protected abstract void startUpClient();
```

Comment:
Seeing as these are optional, you could make these non-abstract empty implementations so the subclasses that don't need them aren't cluttered by them?